src/modules/commander/accelerometer_calibration.cpp

```diff
@@ -370,83 +357,222 @@ int do_accel_calibration(orb_advert_t *mavlink_log_pub)
-				       false) == calibrate_return_ok) {
+	for (unsigned i = 0; i < (sizeof(worker_data.side_data_collected) / sizeof(worker_data.side_data_collected[0])); i++) {
+		if ((cal_mask & (1 << i)) > 0) {
+			// mark as missing
+			worker_data.side_data_collected[i] = false;
+			worker_data.calibration_sides++;
+
+		} else {
+			// mark as completed from the beginning
+			worker_data.side_data_collected[i] = true;
 
-		const Dcmf board_rotation = calibration::GetBoardRotationMatrix();
-		const Dcmf board_rotation_t = board_rotation.transpose();
+			calibration_log_info(mavlink_log_pub,
+					     "[cal] %s side done, rotate to a different side",
+					     detect_orientation_str(static_cast<enum detect_orientation_return>(i)));
+			px4_usleep(100000);
+		}
+	}
+
+	if (calibrate_from_orientation(mavlink_log_pub, worker_data.side_data_collected, accel_calibration_worker,
+				       &worker_data) == calibrate_return_ok) {
+
+		const Rotation board_rotation = calibration::GetBoardRotation();
 
 		bool param_save = false;
 		bool failed = true;
 
-		for (unsigned i = 0; i < MAX_ACCEL_SENS; i++) {
-			if (i < active_sensors) {
-				// calculate offsets
-				Vector3f offset{};
+		int internal_rotation_count{0};
+		Rotation internal_rotations[MAX_ACCEL_SENS] {};
+
+		for (uint8_t cur_accel = 0; cur_accel < active_sensors; cur_accel++) {
+			int accel_data_count = 0;
+
+			matrix::Vector3f accel_data[6] {};
+
+			for (int orientation = 0; orientation < 6; orientation++) {
+				if ((cal_mask & (1 << orientation)) > 0) {
+					accel_data[accel_data_count] = worker_data.accel_ref[cur_accel][orientation];
+					accel_data_count++;
+				}
+			}
+
+			bool sphere_fit_only = false;
+			bool sphere_fit_success = false;
+			bool ellipsoid_fit_success = false;
 
-				// X offset: average X from TAIL_DOWN + NOSE_DOWN
-				const Vector3f accel_tail_down{worker_data.accel_ref[i][ORIENTATION_TAIL_DOWN]};
-				const Vector3f accel_nose_down{worker_data.accel_ref[i][ORIENTATION_NOSE_DOWN]};
-				offset(0) = (accel_tail_down(0) + accel_nose_down(0)) * 0.5f;
+			sphere_params sphere_data{};
+			sphere_data.radius = CONSTANTS_ONE_G;
+			int ret = lm_fit(worker_data.accel_ref[cur_accel], 6, sphere_data, false);
 
-				// Y offset: average Y from LEFT + RIGHT
-				const Vector3f accel_left{worker_data.accel_ref[i][ORIENTATION_LEFT]};
-				const Vector3f accel_right{worker_data.accel_ref[i][ORIENTATION_RIGHT]};
-				offset(1) = (accel_left(1) + accel_right(1)) * 0.5f;
+			if (ret == PX4_OK) {
+				sphere_fit_success = true;
 
-				// Z offset: average Z from UPSIDE_DOWN + RIGHTSIDE_UP
-				const Vector3f accel_upside_down{worker_data.accel_ref[i][ORIENTATION_UPSIDE_DOWN]};
-				const Vector3f accel_rightside_up{worker_data.accel_ref[i][ORIENTATION_RIGHTSIDE_UP]};
-				offset(2) = (accel_upside_down(2) + accel_rightside_up(2)) * 0.5f;
+				if (!sphere_fit_only) {
+					int ellipsoid_ret = lm_fit(accel_data, accel_data_count, sphere_data, true);
```

Comment:
I was thinking of leaving it open for more points in the future, but maybe we should be limiting the fit based on what was actually gathered.

Reply:
So specifically what should we do here? Should we introduce a variation of `lm_ellipsoid_fit_iteration` that skips the off diagonal?

Reply:
In my opinion, adding a new `lm_orthogonal_ellipsoid_fit_iteration` is the best solution.

I have begun to factorize the Levenberg Marquardt before adding a new fit but I have stopped because I was in a rush on other tasks... Some parts of this factorization are not very clean because of the matrix fixed sizes.

I can try to work at this a bit next week and propose a draft pr.

But maybe, in practice, this PR gives good results. I am just afraid that in some conditions it converges to a bad solution....

Reply:
Here is the python code I shared with you on Slack. It is the 6 parameters (3 bias and 3 scales) optimization using gauss newton. Levenberg Marquardt should give equivalent results with potentially faster convergence.

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pyulog
import numpy as np
import matplotlib.pyplot as plt

def sphere_fit( X, Y, Z,norm):

    calib = np.array([1,1,1,0,0,0])

    X_calibGrad = calib[0]*(X-calib[3])
    Y_calibGrad = calib[1]*(Y-calib[4])
    Z_calibGrad = calib[2]*(Z-calib[5])

    delta = 1;
    i=1;

    gain_conv = 0.5;
    maxIter = 100;

    while ((delta>1e-6) and (i<maxIter)):
        prev_calib = calib;

        residuals = X_calibGrad**2 + Y_calibGrad**2 + Z_calibGrad**2 - norm**2

        jac_K1 = 2*calib[0]*(X-calib[3])**2
        jac_K2 = 2*calib[1]*(Y-calib[4])**2
        jac_K3 = 2*calib[2]*(Z-calib[5])**2
        jac_B1 = 2*calib[0]**2*(calib[3]-X)
        jac_B2 = 2*calib[1]**2*(calib[4]-Y)
        jac_B3 = 2*calib[2]**2*(calib[5]-Z)
        J = np.array([jac_K1, jac_K2, jac_K3, jac_B1, jac_B2, jac_B3]).transpose()

        InvJTJ=np.linalg.inv(np.dot(J.transpose(),J));
        calib_error = np.dot(InvJTJ, np.dot(J.transpose(),residuals))

        calib = calib - gain_conv * calib_error

        X_calibGrad = calib[0]*(X-calib[3]);
        Y_calibGrad = calib[1]*(Y-calib[4]);
        Z_calibGrad = calib[2]*(Z-calib[5]);

        delta = sum(abs(calib-prev_calib));
        i=i+1;

    print('nb iter ' + str(i))
    print('delta ' + str(delta))

    return calib

def plot_sphere(r):

    # Create a sphere
    r = g
    pi = np.pi
    cos = np.cos
    sin = np.sin
    phi, theta = np.mgrid[0.0:pi:100j, 0.0:2.0*pi:100j]
    x = r*sin(phi)*cos(theta)
    y = r*sin(phi)*sin(theta)
    z = r*cos(phi)

    #Import data
    #Set colours and render
    ax = fig.add_subplot(111, projection='3d')
    ax.plot_surface(
        x, y, z,  rstride=1, cstride=1, color='c', alpha=0.6, linewidth=0)

    ax.set_box_aspect(aspect = (1,1,1))


g = 9.80665
log = pyulog.ULog('path_to_the_log_file.ulg')
data = {log.data_list[i].name + '_' + str(log.data_list[i].multi_id) : log.data_list[i].data
          for i in range(0, len(log.data_list))}

print(data['sensor_accel_0']['device_id'][0])
calib0 = sphere_fit(data['sensor_accel_0']['x'],
                    data['sensor_accel_0']['y'],
                    data['sensor_accel_0']['z'],
                    g)

print(calib0)
print()
fig = plt.figure()
plot_sphere(g)
plt.plot(data['sensor_accel_0']['x'],
         data['sensor_accel_0']['y'],
         data['sensor_accel_0']['z'],
         'b')
plt.plot(calib0[0]*(data['sensor_accel_0']['x']-calib0[3]),
         calib0[1]*(data['sensor_accel_0']['y']-calib0[4]),
         calib0[2]*(data['sensor_accel_0']['z']-calib0[5]),
         'r')
plt.tight_layout()


calib1 = sphere_fit(data['sensor_accel_1']['x'],
                    data['sensor_accel_1']['y'],
                    data['sensor_accel_1']['z'],
                    g)

calib2 = sphere_fit(data['sensor_accel_2']['x'],
                    data['sensor_accel_2']['y'],
                    data['sensor_accel_2']['z'],
                    g)


print(data['sensor_accel_1']['device_id'][0])
print(calib1)
print()
print(data['sensor_accel_2']['device_id'][0])
print(calib2)
```